Remove one of the two worklists from the SIL linker

slavapestov opened the pull request:
I'm re-doing the control flow here in preparation for landing a change making the linker lazier in -Onone.

slavapestov committed:
SIL Linker: Factor out maybeAddFunctionToWorklist()

The check to see if a function should be added to the worklist was
repeated in a few places.

slavapestov committed:
SIL Linker: Remove one of the two worklists

The second worklist was immediately drained after visiting each

SIL instruction, so it didn't really avoid any recursion.
Simplify the control flow by immediately deserializing the
function and then adding it to the primary worklist.

slavapestov committed:
SIL Linker: Clean up processFunction()

Now that addFunctionToWorklist() immediately deserializes,
we can re-use it here.

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please test source compatibility

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please test source compatibility

slavapestov commented:
@swift-ci Please test source compatibility

slavapestov commented:
@swift-ci Please test source compatibility

gottesmm commented on lib/SIL/Linker.h:
```diff
@@ -79,10 +79,9 @@ class SILLinkerVisitor : public SILInstructionVisitor<SILLinkerVisitor, bool> {
-  }
+  bool addFunctionToWorklist(SILFunction *F);
+
+  bool maybeAddFunctionToWorklist(SILFunction *F);
```
Please add doxygen comments here.

gottesmm commented on lib/SIL/Linker.cpp:
```diff
@@ -70,10 +85,7 @@ bool SILLinkerVisitor::linkInVTable(ClassDecl *D) {
-      Result = true;
-      addFunctionToWorklist(P.Implementation);
-    }
+    Result |= maybeAddFunctionToWorklist(P.Implementation);
```
I haven't thought about this (I am asking you to do so), but are you sure that having a uniform condition across all of these is safe. For instance just off the top of my head, imagine if we could specialize witness tables. Would the linkage be shared? If we then serialized that, wouldn't we have shared external linkage and then wouldn't deserialize the function?

Also given that you are changing the conditions, can you add tests that verify this new behavior for other parts of this patch?

gottesmm replied:
Safe is the wrong word here, I meant are you sure that expanding the conditions here what we want to do.

slavapestov replied:
hasSharedVisibility() returns true for both shared and shared external. Also note that linkInVTable() is only called it isLinkAll() is true, so in this case the condition is equivalent to F->isExternalDefinition(). So there’s no real functional change here. I’ll add an assert and comments.

gottesmm commented on lib/SIL/Linker.cpp:
```diff
@@ -218,7 +219,7 @@ bool SILLinkerVisitor::visitProtocolConformance(
       // to deserialize.
       performFuncDeserialization = true;
-      addFunctionToWorklist(E.getMethodWitness().Witness);
+      maybeAddFunctionToWorklist(E.getMethodWitness().Witness);
```
You are also expanding the conditions here. Same concern: what about specialized witness tables? That being said, maybe this is not interesting.

slavapestov replied:
The only way we can end here in the mandatory pipeline (where isLinkAll is off) is for imported conformances, where all witnesses are shared. So again there’s no real change here. We are just avoiding adding functions to the work list if they already have a body.

gottesmm reviewed (commented):
Reviewed the first patch. I need to look at the other ones in context of the branch to understand.

slavapestov commented:
@gottesmm Should we rename LinkNormal to LinkMandstory and LinkAll to LinkPerformance?

slavapestov committed:
SIL: Add SILModule::loadFunction()

This is like linkFunction(), except it only deserializes the immediate
function body, without recursively walking its transitive dependencies.

slavapestov committed:
SIL Linker: Use SILModule::loadFunction()

This means SILLinkerVisitor no longer depends on the SILLoader;
the latter is now just an implementation detail of SILModule.

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please test source compatibility

slavapestov commented:
@swift-ci Please test source compatibility

slavapestov commented:
@swift-ci Please test source compatibility

slavapestov commented:
@swift-ci Please test source compatibility
